shiny/ui/_theme.py

```diff
@@ -0,0 +1,513 @@
+
+    @classmethod
+    def available_presets(cls) -> tuple[ShinyThemePreset, ...]:
+        """
+        Get a list of available theme presets.
+        """
+        return shiny_theme_presets
+
+    @property
+    def preset(self) -> ShinyThemePreset:
+        return self._preset
+
+    @preset.setter
+    def preset(self, value: ShinyThemePreset) -> None:
+        check_is_valid_preset(value)
+        self._preset = value
+        self._reset_css()
+
+    def _reset_css(self) -> None:
+        self._css = ""
+        self._css_temp_srcdir = None
+
+    def _has_customizations(self) -> bool:
+        return (
+            len(self._functions) > 0
+            or len(self._defaults) > 0
+            or len(self._mixins) > 0
+            or len(self._rules) > 0
+        )
+
+    def _can_use_precompiled(self) -> bool:
+        return (
+            self._preset in shiny_theme_presets_bundled
+            and not self._has_customizations()
+        )
+
+    def _combine_args_kwargs(
+        self,
+        *args: str,
+        allow_both: bool = True,
+        kwargs: dict[str, str | float | int | bool | None],
+        is_default: bool = False,
+    ) -> list[str]:
+        if not allow_both and len(args) > 0 and len(kwargs) > 0:
+            # Python forces positional arguments to come _before_ kwargs, but default
+            # argument order might matter. To be safe, we force users to pick one order.
+            raise ValueError("Cannot provide both positional and keyword arguments.")
+        if len(args) == 0 and len(kwargs) == 0:
+            return []
+
+        values: list[str] = list(args)
+        default = " !default" if is_default else ""
+
+        if len(kwargs) > 0:
+            for key, value in kwargs.items():
+                key = key.replace("_", "-")
+                if isinstance(value, bool):
+                    value = "true" if value else "false"
+                elif value is None:
+                    value = "null"
+                values.append(f"${key}: {value}{default};")
+
+        return dedent_array(values)
+
+    def add_functions(
+        self: T,
+        *args: str,
+        **kwargs: str | float | int | bool | None,
+    ) -> T:
+        """
+        Add custom Sass functions to the theme.
+
+        Sass code added via this method will be placed **after** the function
+        declarations from the theme preset, allowing you to override or extend the
+        default functions.
+
+        Parameters
+        ----------
+        *args
+            The Sass functions to add as a single or multiple strings.
+        **kwargs
+            Keyword arguments containing function declarations to add. The keys
+            should be Sass variable names using underscore casing that will be
+            transformed automatically to kebab-case. For example,
+            `.add_functions(primary_color="#ff0000")` is equivalent to
+            `.add_functions("$primary-color: #ff0000;")`.
```

Comment:
If examples are provided, make them relevant to functions

Reply:
And actually, I think I'd prefer not to have `kwargs` on `add_functions()`? I can see why having them on mixins and rules could be useful, but it seems like using them in `add_functions()` would be an anti-pattern (i.e., they should go in `.add_defaults()` instead?